### PR TITLE
feat: add UI For Auth HTTP Hooks

### PR DESCRIPTION
**Pull request by J0**
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.
YES

An example SQL function for testing the Postgres function as a hook feature.
Screenshot 2024-05-27 at 14 29 42

### CODE REVIEW

File: packages/ui/src/components/Toggle/Toggle.tsx [hunk 1]
```diff
@@ -48,7 +48,7 @@ function Toggle({
 }: Props) {
   const __styles = styleHandler('toggle')
 
-  const { formContextOnChange, values, errors, handleBlur, touched, fieldLevelValidation } =
+  const { values, errors, handleBlur, touched, fieldLevelValidation, setFieldValue } =
```
Comment (MildTomato) [inline packages/ui/src/components/Toggle/Toggle.tsx:51]: woah. what's going on

File: packages/ui/src/components/Toggle/Toggle.tsx [hunk 2]
```diff
@@ -83,24 +83,8 @@ function Toggle({
 
     setIntChecked(!intChecked)
 
-    /*
-     * Create change event for formik
-     * formik expects an input change event
-     */
-    let event: any = {}
```
Comment (MildTomato) [inline packages/ui/src/components/Toggle/Toggle.tsx:90]: this will break something

**Review by MildTomato — CHANGES_REQUESTED** [review]
@supabase/frontend can we move to using new components and using Forms better in this new Auth hooks UI.

**Comment by mu4farooqi** [comment]
Can we use custom SMS provider webhooks without this PR being merged?

Like with CLI?

**Comment by J0** [comment]
Not yet, until we merge the CLI bindings which require this PR to be merged as well as the docs which also depend on the PR.


Sorry about that - we are trying to push forward though.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                      label="Schema"
+                      className="px-8"
+                      description="Postgres schema where the function is defined."
+                    >
+                      <FormControl_Shadcn_>
+                        <SchemaSelector
+                          size="small"
+                          showError={false}
+                          selectedSchemaName={field.value}
+                          disabled={field.disabled}
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:253]: you don't need to do this

just `{...field}`

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+                      label="Function name"
+                      className="px-8"
+                      description="Postgres function which will be called by Supabase Auth each time the hook is triggered."
+                    >
+                      <FormControl_Shadcn_>
+                        <FunctionSelector
+                          size="small"
+                          schema={values.postgresValues.schema}
+                          selectedFunctionName={field.value}
+                          onSelectFunction={(name) => field.onChange(name)}
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:277]: can we rename these props to just be standard?

File: packages/ui/src/components/Toggle/Toggle.tsx (hunk 1 above)
Comment (ivasilov) [reply packages/ui/src/components/Toggle/Toggle.tsx]: Change reverted.

File: packages/ui/src/components/Toggle/Toggle.tsx (hunk 2 above)
Comment (ivasilov) [reply packages/ui/src/components/Toggle/Toggle.tsx]: Reverted change.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx (hunk 3 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:338]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx (hunk 4 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+              />
+            </div>
+            <Separator />
+            <FormField_Shadcn_
+              control={form.control}
+              name="selectedType"
+              render={({ field }) => (
+                <FormItemLayout label="Hook type" className="px-8">
+                  <FormControl_Shadcn_>
+                    <Radio.Group
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:277]: old component

**Review by MildTomato — DISMISSED** [review]

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            <DropdownMenuTrigger asChild>
+              <Button
+                type="default"
+                iconRight={<ChevronDown className="w-4 h-4" strokeWidth={1} />}
+              >
+                Add hook
+              </Button>
+            </DropdownMenuTrigger>
+            <DropdownMenuContent className="w-76 p-0" align="end">
+              <div className="p-2">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:86]: no need for padding

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+          <DropdownMenu modal={false}>
+            <DropdownMenuTrigger asChild>
+              <Button
+                type="default"
+                iconRight={<ChevronDown className="w-4 h-4" strokeWidth={1} />}
+              >
+                Add hook
+              </Button>
+            </DropdownMenuTrigger>
+            <DropdownMenuContent className="w-76 p-0" align="end">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:85]: why `p-0` ?

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx (hunk 6 above)
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:86]: `p-1`. DropdownContent uses `p-1` which I'm guessing it what your emulating

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+              <div className="p-2">
+                {nonEnterpriseHookOptions.map((h) => (
+                  <DropdownMenuItem key={h.title} onClick={() => setSelectedHook(h.title)}>
+                    {h.title}
+                  </DropdownMenuItem>
+                ))}
+              </div>
+              {nonEnterpriseHookOptions.length > 0 && <DropdownMenuSeparator />}
+
+              <div className="bg-surface-200 p-2 -mt-2">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:95]: same, `p-1`

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+interface HookCardProps {
+  hook: Hook
+  canUpdateConfig: boolean
+  onToggle: (enabled: boolean) => void
+  onSelect: () => void
+}
+
+export const HookCard = ({ hook, canUpdateConfig, onToggle, onSelect }: HookCardProps) => {
+  return (
+    <div className="bg-surface-100 border-default overflow-hidden border shadow px-5 py-4 flex flex-row">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:14]: `first:rounded-t-md last:rounded-b-md`

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+      <div className="px-3 pt-1">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:15]: Maybe don't use paddin here.
Use padding on container div above and use gap better.

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        <div className="text-sm flex flex-row space-x-5 py-4">
+          {hook.method.type === 'postgres' ? (
+            <div className="flex flex-col w-full space-y-2">
+              <div className="flex flex-row items-center">
+                <span className="text-foreground-light w-20">type</span>
+                <span className="text-foreground">Postgres function</span>
+              </div>
+              <div className="flex flex-row items-center">
+                <span className="text-foreground-light w-20">schema</span>
+                <Input
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:30]: we need new components for these

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        </div>
+        <div className="flex flex-row gap-2">
+          <Button type="default" disabled={!canUpdateConfig} onClick={() => onSelect()}>
+            Configure hook
+          </Button>
+        </div>
+      </div>
+      <div className="p-4 flex-1">
+        {hook.enabled ? (
+          <Badge className="space-x-1" size="large" variant="brand">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:94]: Screenshot 2024-05-21 at 11 01 21 AM

Need to add another Badge variant I think.

File: apps/studio/pages/project/[ref]/auth/hooks.tsx [hunk 13]
```diff
@@ -16,12 +14,9 @@ const Hooks: NextPageWithLayout = () => {
     return <NoPermission isFullPage resourceText="access your project's auth hooks" />
   } else {
     return (
-      <FormsContainer>
-        <div className="flex flex-col gap-8">
-          <BasicHooksConfig />
-          <EnterpriseHooksConfig />
-        </div>
-      </FormsContainer>
+      <div className="mx-auto max-w-4xl px-5 pt-12 pb-20">
```
Comment (MildTomato) [inline apps/studio/pages/project/[ref]/auth/hooks.tsx:17]: Scaffold component ?

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+  canUpdateConfig: boolean
+  onToggle: (enabled: boolean) => void
+  onSelect: () => void
+}
+
+export const HookCard = ({ hook, canUpdateConfig, onToggle, onSelect }: HookCardProps) => {
+  return (
+    <div className="bg-surface-100 border-default overflow-hidden border shadow px-5 py-4 flex flex-row">
+      <div className="px-3 pt-1">
+        <Webhook size={24} strokeWidth="0.5" />
```
[the remark this answers is not in the document]
Comment (MildTomato) [reply apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:16]: is it not 1, with 21x21 size ?

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+                  >
+                    {h.title}
+                  </DropdownMenuItem>
+                ))}
+              </div>
+            </DropdownMenuContent>
+          </DropdownMenu>
+        }
+      />
+      <div className="first:rounded-t-md last:rounded-b-md pb-2">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:127]: `first:rounded-t-md last:rounded-b-md` you need these on the cards themselves.
unless you did 

`[&>div]:first:rounded-t-md [&>div]:last:rounded-b-md` is probably what you wanted here.

i've added the comments to HookCard.tsx

Comment (MildTomato) [reply apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:127]: you will need `-space-y-px` on this div btw, so you don't get the double lines.
Screenshot 2024-05-21 at 11 12 04 AM

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx (hunk 5 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx (hunk 10 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:15]: This is how it looks without top padding:
Screenshot 2024-05-27 at 13 32 09

A bit weird to me, the icon being higher than the text.

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx (hunk 6 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx]: FIxed.

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx (hunk 8 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx (hunk 9 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx (hunk 15 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx (hunk 11 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/HookCard.tsx:30]: You probably had an old branch, this is how they look now.

Screenshot 2024-05-27 at 13 53 50

File: apps/studio/pages/project/[ref]/auth/hooks.tsx (hunk 13 above)
Comment (ivasilov) [reply apps/studio/pages/project/[ref]/auth/hooks.tsx]: Fixed.

**Comment by saltcod** [comment]
are there any docs? or going to be a blog post or anything? Something to link to with a "Learn more"

screenshot-2024-05-27-at-12 36 49

**Comment by saltcod** [comment]
can I disable this? Would I ever want to? 

screenshot-2024-05-27-at-12 55 20

**Comment by saltcod** [comment]
I can delete it, nm!

**Review by saltcod — APPROVED** [review]

File: apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+      ),
+    }
+  })
+
+  const nonEnterpriseHookOptions = hooks.filter((h) => !isValidHook(h) && !h.enterprise)
+  const enterpriseHookOptions = hooks.filter((h) => !isValidHook(h) && h.enterprise)
+
+  return (
+    <>
+      <FormHeader
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/HooksListing.tsx:72]: Can we use `<Scaffold>` ?

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 17]
```diff
@@ -0,0 +1,425 @@
...
+          </div>
+        </SheetHeader>
+        <Separator />
+        <Form_Shadcn_ {...form}>
+          <form
+            id={FORM_ID}
+            className="space-y-6 w-full py-8 flex-1"
+            onSubmit={form.handleSubmit(onSubmit)}
+          >
+            <div>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:254]: random div?

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 18]
```diff
@@ -0,0 +1,425 @@
...
+                      <FormControl_Shadcn_>
+                        <div className="flex flex-row">
+                          <Input_Shadcn_ {...field} className="rounded-r-none border-r-0" />
+                          <Button
+                            type="default"
+                            size="small"
+                            className="rounded-l-none h-[38px]"
+                            onClick={() => {
+                              const authHookSecret = generateAuthHookSecret()
+                              field.onChange(authHookSecret)
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:387]: not sure it makes any difference but you could use 

`form.setValue('httpsValues.secret', authHookSecret)` 

just so you don't hijack the field.onChange as you'd expect that to come from the input, not this button.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 19]
```diff
@@ -0,0 +1,425 @@
...
+          >
+            <div>
+              <FormField_Shadcn_
+                key="enabled"
+                control={form.control}
+                name="enabled"
+                render={({ field }) => (
+                  <FormItem_Shadcn_ className="px-8">
+                    <FormControl_Shadcn_>
+                      <Toggle
```
Comment (MildTomato) [inline apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:262]: pleas use `<Switch/>`

think I better get the deprecation notices in.

Comment (MildTomato) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:262]: see link above, shows Switch used in a FormItemLayout component

Comment (MildTomato) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx:262]: https://github.com/supabase/supabase/pull/26817
deprecation notices PR

Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx (hunk 17 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx [hunk 20]
```diff
@@ -0,0 +1,425 @@
...
+  return `v1,whsec_${base64String}`
+}
+
+const FORM_ID = 'create-edit-auth-hook'
+
+const FormSchema = z
+  .object({
+    hookType: z.string(),
+    enabled: z.boolean(),
+    selectedType: z.string(),
```
[the remark this answers is not in the document]
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

File: apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx (hunk 18 above)
Comment (ivasilov) [reply apps/studio/components/interfaces/Auth/Hooks/CreateHookSheet.tsx]: Fixed.

**Comment by ivasilov** [comment]
> are there any docs? or going to be a blog post or anything? Something to link to with a "Learn more"

There's no docs at the moment. I'll ask @J0 to add a link when he adds the docs.

**Pull request merged**